src/home_robot/home_robot/mapping/voxel_map.py

```diff
@@ -2,25 +2,329 @@
+
+    def sample_frontier(
+        self,
+        max_tries_per_size: int = 100,
+        min_size: int = 5,
+        max_size: int = 10,
+        debug: bool = False,
+        verbose: bool = False,
+    ) -> Optional[torch.Tensor]:
+        """Sample a valid location on the current frontier. Works by finding the edges of "explored" that are not obstacles.
+
+        Args:
+            max_tries_per_size(int): number for rejection sampling
+            min_size(int): min radius of filter for growing frontier
+            max_size(int): max radius of filter for growing frontier
+            debug(bool): show visualizations of frontiers
+        """
+
+        # Get the masks from our 3d map
+        obstacles, explored = self.voxel_map.get_2d_map()
+
+        # Extract edges from our explored mask
+        less_explored = binary_erosion(
+            explored.float().unsqueeze(0).unsqueeze(0), self.dilate_explored_kernel
+        )[0, 0]
+        edges = get_edges(less_explored)
+
+        # Do not explore obstacles any more
+        frontier_edges = edges & ~obstacles
+
+        for radius in range(min_size, max_size + 1):
+            # Now we apply this filter and try to sample a goal position
+            if verbose:
+                print("[VOXEL MAP: sampling] sampling margin of size", radius)
+            expanded_frontier = expand_mask(frontier_edges, radius)
+            # TODO: should we do this or not?
+            # Make sure not to sample things that will just be in obstacles
+            # expanded_obstacles = expand_mask(obstacles, radius)
+
+            # Mask where we will look at
+            outside_frontier = ~explored & ~obstacles
+
+            # Mask where we will sample locations to move to
+            expanded_frontier = expanded_frontier & explored & ~obstacles
+
+            if debug:
+                import matplotlib.pyplot as plt
+
+                plt.subplot(221)
+                plt.imshow(frontier_edges.cpu().numpy())
+                plt.subplot(222)
+                plt.imshow(expanded_frontier.cpu().numpy())
+                plt.title("expanded frontier")
+                plt.subplot(223)
+                plt.imshow(outside_frontier.cpu().numpy())
+                plt.title("outside frontier")
+                plt.subplot(224)
+                plt.imshow((less_explored + explored).cpu().numpy())
+                plt.title("explored")
+                plt.show()
+
+            # TODO: this really should not be random at all
+            valid_indices = torch.nonzero(expanded_frontier, as_tuple=False)
+            if valid_indices.size(0) == 0:
+                continue
+
+            # Rejection sampling:
+            # - Find a point that we could potentially move to
+            # - Compute a position and orientation
+            # - Check to see if we can actually move there
+            # - If so, return it
+            for i in range(max_tries_per_size):
```

Comment:
I think this is reasonable and it's good to have a first version that works. 
If RRT becomes slow then this rejection sampling should be easy to parallelize/run on GPU

Reply:
Yeah imo this is super easy to parallelize, just not something i wanted to take on just yet
will probably be a good way to guarantee reachable placement poses later though